File: find_the_duplicate_number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int slow = 0, fast = 0;
        do {
            slow = nums[slow];
            fast = nums[nums[fast]];
        } while(slow != fast);

        slow = 0;

        while(slow != fast) {
            fast = nums[fast];
            slow = nums[slow];
        }
        return fast;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(1)

File: find_the_duplicate_number.go
func findDuplicate(nums []int) int {
    slow, fast := 0, 0

    for {
        slow = nums[slow]
        fast = nums[nums[fast]]

        if nums[slow] == nums[fast] {
            break
        }
    }

    slow = 0
    for slow != fast {
        slow = nums[slow]
        fast = nums[fast]
    }
    return fast
}

// Time Complexity : O(n)
// Space Complexity : O(1)